File: mobile/src/main/java/com/bayapps/android/robophish/model/ParseUtils.java
package com.bayapps.android.robophish.model;

import org.json.JSONArray;
import org.json.JSONObject;

import java.text.SimpleDateFormat;
import java.util.Date;

import timber.log.Timber;

/**
 * Created by mvoytovich on 1/13/15.
 */
public class ParseUtils {

    public static Show parseShow(JSONObject json) {

        Show show;

        try {

            //parse 'data'
            JSONObject data = json.getJSONObject("data");
            show =  parseShowData(data);

        } catch (Exception e) {
            Timber.d("failed to parse show!");
            e.printStackTrace();
            return null;
        }

        return show;
    }

    private static Show parseShowData(JSONObject data) {

        JSONArray tracks;

        Show show = new Show();

        try {

            //parse 'data'
            Long showId = data.getLong("id");
            show.setId(showId);

            String dateString = data.getString("date"); //formatted as YYYY-MM-DD
            Date date = new SimpleDateFormat("yyyy-MM-dd").parse(dateString);
            show.setDate(date);

            JSONObject venue = data.optJSONObject("venue");
            if (venue != null) {
                //full show data contains a venue object
                String venueName = venue.getString("name");
                show.setVenueName(venueName);

                String location = venue.getString("location");
                show.setLocation(location);

            } else {
                String venueName = data.getString("venue_name");
                show.setVenueName(venueName);

                String location = data.getString("location");
                show.setLocation(location);
            }

            //parse 'tracks' if available (if this is more that "simple data")
            tracks = data.optJSONArray("tracks");

            if (tracks != null) {
                for (int i = 0; i < tracks.length(); i++) {
                    JSONObject jsonTrack = tracks.getJSONObject(i);
                    Timber.d(jsonTrack.toString());

                    int id = jsonTrack.getInt("id");
                    String title = jsonTrack.getString("title");
                    String url = jsonTrack.getString("mp3");

                    long duration = jsonTrack.getLong("duration");

                    Track track = new Track(id, title, url);
                    track.setDuration(duration);

                    show.addTrack(track);
                }
            }

            String taperNotes = data.getString("taper_notes");
            if (taperNotes != null) {
                show.setTaperNotes(taperNotes);
            }

            boolean sbd = data.getBoolean("sbd");
            show.setSbd(sbd);

        } catch (Exception e) {
            Timber.d("failed to parse show!");
            e.printStackTrace();
            return null;
        }

        return show;
    }
}

File: mobile/src/main/java/com/bayapps/android/robophish/model/ParseUtils.kt
package com.bayapps.android.robophish.model

import org.json.JSONObject
import timber.log.Timber
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

object ParseUtils {
    fun parseShow(json: JSONObject): Show? {
        return runCatching {
            val data = json.getJSONObject("data")
            parseShowData(data)
        }.getOrElse {
            Timber.d("failed to parse show!")
            Timber.e(it)
            null
        }
    }

    private fun parseShowData(data: JSONObject): Show? {
        return runCatching {
            val show = Show()
            show.id = data.getLong("id")

            val dateString = data.getString("date")
            val date = SimpleDateFormat("yyyy-MM-dd", Locale.US).parse(dateString)
            show.date = date

            val venue = data.optJSONObject("venue")
            if (venue != null) {
                show.venueName = venue.getString("name")
                show.location = venue.getString("location")
            } else {
                show.venueName = data.getString("venue_name")
                show.location = data.getString("location")
            }

            val tracks = data.optJSONArray("tracks")
            if (tracks != null) {
                for (i in 0 until tracks.length()) {
                    val jsonTrack = tracks.getJSONObject(i)
                    Timber.d(jsonTrack.toString())

                    val id = jsonTrack.getInt("id")
                    val title = jsonTrack.getString("title")
                    val url = jsonTrack.getString("mp3")
                    val duration = jsonTrack.getLong("duration")

                    val track = Track(id.toLong(), title, url)
                    track.duration = duration
                    show.addTrack(track)
                }
            }

            val taperNotes = data.optString("taper_notes", null)
            show.taperNotes = taperNotes

            show.isSbd = data.getBoolean("sbd")
            show
        }.getOrElse {
            Timber.d("failed to parse show!")
            Timber.e(it)
            null
        }
    }
}
